### PR TITLE
fix omp scan bug

### LINKED ISSUE #6317: [BUG]: stdpar multicore scan tests fail at runtime after commit 1238f287f4f654bff12e4d4a29b7dfa1c40a12fe
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Runtime Error

### Component

Thrust

### Describe the bug

stdpar multicore scan tests have started failing in correctness after the following commit: 1238f287f4f654bff12e4d4a29b7dfa1c40a12fe
For this example:
```
#include <execution>
#include <algorithm>
#include <numeric>
#include <cassert>

constexpr int N = 10;

int main() {
  int* data = new int[N];
  int* output = new int[N];
  std::iota(data, data + N, 0);
  auto r0 = std::exclusive_scan(std::execution::par, data, data + N, output, 10);
  assert(r0 == output + N);
  delete[] output;
  delete[] data;
}
```
The program fails at runtime with the following assertion error:
```
int main(): Assertion `r0 == output + N' failed.
```

### How to Reproduce

```
nvc++ -stdpar=multicore -fast --c++17 scan.cpp
```

### Expected behavior

The program should pass the assertion (r0 == output + N) and complete successfully.

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by charan-003**
using the first element of each block as the init value

closes #6317

### CODE REVIEW

File: thrust/thrust/system/omp/detail/scan.h [hunk 1]
```diff
@@ -106,7 +106,25 @@ OutputIterator scan_impl(
...
+        }
+      }
+      else
+      {
+        // No init value: reduce uses first element as init
+        if (end > start)
+        {
+          block_sums[tid] = ::cuda::std::reduce(first + start + 1, first + end, *(first + start), wrapped_binary_op);
+        }
+      }
```
Comment (dkolsen-pgi) [inline thrust/thrust/system/omp/detail/scan.h]: You want to always do the `else` branch.  The true branch is incorrect.  The initial value doesn't get incorporated into the result until step 2.  And you still have `accum_t{}` in there, which is not correct.  Get rid of the `if` statement entirely and just replace
```c++
block_sums[tid] = ::cuda::std::reduce(first + start, first + end, accum_t{}, wrapped_binary_op);
```
with
```c++
block_sums[tid] = ::cuda::std::reduce(first + start + 1, first + end, *(first + start), wrapped_binary_op);
```

File: thrust/thrust/system/omp/detail/scan.h [hunk 2]
```diff
@@ -106,7 +106,25 @@ OutputIterator scan_impl(
...
+        }
+        else
+        {
+          block_sums[tid] = ::cuda::std::reduce(first + start, first + end, accum_t{}, wrapped_binary_op);
+        }
+      }
+      else
+      {
+        // No init value: reduce uses first element as init
+        if (end > start)
```
Comment (dkolsen-pgi) [inline thrust/thrust/system/omp/detail/scan.h]: This check is unnecessary.  If `start < n`, then `end > start` is always true.

File: thrust/thrust/system/omp/detail/scan.h [hunk 3]
```diff
@@ -117,7 +135,18 @@ OutputIterator scan_impl(
...
+    {
+      temporary_array<accum_t, DerivedPolicy> block_sums_copy(exec, num_threads);
+      for (int i = 0; i < num_threads; ++i)
+      {
+        block_sums_copy[i] = block_sums[i];
+      }
+      ::cuda::std::inclusive_scan(
+        block_sums_copy.begin(), block_sums_copy.begin() + num_threads - 1, block_sums.begin() + 1, wrapped_binary_op);
+      // block_sums[0] stays as-is (no prefix for first block)
+    }
```
Comment (dkolsen-pgi) [inline thrust/thrust/system/omp/detail/scan.h]: This is more complex than necessary.  There is no need for `block_sums_copy`.  For the inclusive_scan with no init, just do an exclusive_scan starting at the second element where the initial value is the first element.
```
::cuda::std::exclusive_scan(block_sums.begin() + 1, block_sums.end(), block_sums.begin() + 1,
                            block_sums[0], wrapped_binary_op);
```

File: thrust/thrust/system/omp/detail/scan.h [hunk 4]
```diff
@@ -130,13 +159,30 @@ OutputIterator scan_impl(
...
+          if (tid == 0)
+          {
+            ::cuda::std::inclusive_scan(first + start, first + end, result + start, wrapped_binary_op);
+          }
+          else
+          {
+            const accum_t prefix = block_sums[tid];
+            ::cuda::std::inclusive_scan(first + start, first + end, result + start, wrapped_binary_op, prefix);
+          }
+        }
```
Comment (dkolsen-pgi) [inline thrust/thrust/system/omp/detail/scan.h:160]: This part of the change looks correct to me.  There is no way to avoid the extra branching.

**Review by dkolsen-pgi — CHANGES_REQUESTED** [review]

**Comment by bernhardmgruber** [comment]
@charan-003 I added a few more tests.

**Comment by bernhardmgruber** [comment]
@charan-003 I will review this next week.

File: thrust/testing/transform_scan.cu [hunk 5]
```diff
@@ -400,3 +403,145 @@ void TestValueCategoryDeduction()
...
+void TestTransformScanEdgeCases()
+{
+  // Test 1: Large array with transform_inclusive_scan, multiplies, and init
+  {
+    const int n = 10000;
+    thrust::device_vector<int> d_input(n);
+    for (int i = 0; i < n; ++i)
+    {
+      d_input[i] = (i % 5) + 1;
+    }
```
Comment (bernhardmgruber) [inline thrust/testing/transform_scan.cu]: Important: This is **really** slow, since this will do a `cudaMemcpyAsync` for each subscription of `d_input`, so 10000 times. Please use `thrust::tabulate` and a functor to initialize the data.

File: thrust/testing/transform_scan.cu [hunk 6]
```diff
@@ -400,3 +403,145 @@ void TestValueCategoryDeduction()
...
+  }
+
+  // Test 2: transform_exclusive_scan with multiplies on large array
+  {
+    const int n = 10000;
+    thrust::device_vector<int> d_input(n);
+    for (int i = 0; i < n; ++i)
+    {
+      d_input[i] = (i % 3) + 1;
+    }
```
Comment (bernhardmgruber) [inline thrust/testing/transform_scan.cu]: Same here. Applies to more places (also in the other test file)

File: thrust/thrust/system/omp/detail/scan.h [hunk 7]
```diff
@@ -106,7 +106,9 @@ OutputIterator scan_impl(
 
     if (start < n)
     {
-      block_sums[tid] = ::cuda::std::reduce(first + start, first + end, accum_t{}, wrapped_binary_op);
+      // For both has_init and no-init cases: reduce each block using first element as init
+      accum_t first_elem = *(first + start);
+      block_sums[tid]    = ::cuda::std::reduce(first + start + 1, first + end, first_elem, wrapped_binary_op);
```
Comment (bernhardmgruber) [inline thrust/thrust/system/omp/detail/scan.h]: Important: Let's avoid an extra copy:
```suggestion
      block_sums[tid]    = ::cuda::std::reduce(first + start + 1, first + end, *(first + start), wrapped_binary_op);
```

Comment (bernhardmgruber) [inline thrust/thrust/system/omp/detail/scan.h:114]: I am wondering what happens if `num_threads` becomes larger than `parallel_scan_threshold`. In this case `block_size` becomes zero and `start` and `end` are the same. I think we should catch this sooner where we have:
```c++
if (static_cast<size_t>(n) < parallel_scan_threshold || num_threads <= 1)
```
and do:
```c++
if (static_cast<size_t>(n) < ::cuda::std::max(parallel_scan_threshold, num_threads) || num_threads <= 1)
```

File: thrust/thrust/system/omp/detail/scan.h [hunk 8]
```diff
@@ -117,7 +119,13 @@ OutputIterator scan_impl(
   }
   else
   {
-    ::cuda::std::exclusive_scan(block_sums.begin(), block_sums.end(), block_sums.begin(), accum_t{}, wrapped_binary_op);
+    // For no init inclusive scan: exclusive_scan starting at second element with first element as init
+    if (num_threads > 1)
```
Comment (bernhardmgruber) [inline thrust/thrust/system/omp/detail/scan.h]: I think this condition is always true, since we handle `num_threads <= 1` serially above. I think we can remove the `if` and just use the body directory. You can of course replace it by an assertion instead:
```suggestion
    _CCCL_ASSERT (num_threads > 1, "");
```

**Comment by charan-003** [comment]
@bernhardmgruber I was wondering if it's possible to perform CI tests locally without using Docker?

**Comment by bernhardmgruber** [comment]
> @bernhardmgruber I was wondering if it's possible to perform CI tests locally without using Docker?

Sure! CCCL is just an ordinary cmake project so you can:
```
mkdir my_build
cd my_build
cmake .. -DTHRUST_ENABLE_MULTICONFIG=OFF
ccmake . # enable what you need, for example Thrust. make sure to set the THRUST_DEVICE_SYSTEM=OMP
make thrust.test.scan
./bin/thrust.test.scan
```

File: thrust/testing/transform_scan.cu [hunk 9]
```diff
@@ -400,3 +404,142 @@ void TestValueCategoryDeduction()
...
+    ASSERT_EQUAL(vec, expected);
+  }
+
+  // Test 5: Boundary case at threshold (1024 elements)
+  {
+    const int n = 1024;
+    thrust::device_vector<int> d_input(n);
+    thrust::tabulate(d_input.begin(), d_input.end(), [](int i) {
+      return i + 1;
+    });
```
Comment (bernhardmgruber) [inline thrust/testing/transform_scan.cu]: I am not sure we enable the use of lambdas in tests by default. We have a few options:
1. turn the lambdas into function objects
2. use `thrust::placeholders` like:
```suggestion
    using thrust::placeholders::_1;
    thrust::tabulate(d_input.begin(), d_input.end(), _1 + 1);
```
3. add an equally named `transform_scan.cmake` file with the content:
```
target_compile_options(
  ${test_target}
  PRIVATE $<$<COMPILE_LANG_AND_ID:CUDA,NVIDIA>:--extended-lambda>
)
```
I am leaning towards 2 or 3.

Comment (charan-003) [reply thrust/testing/transform_scan.cu]: sure, i will be using `thrust::placeholders`.
let me make those changes asap 
thank you :))

**Review by bernhardmgruber — APPROVED** [review]

**Comment by bernhardmgruber** [comment]
Thank you so much for taking the time to push the fix through!

**Comment by bernhardmgruber** [comment]
@dkolsen-pgi please re-review and unblock. Thx!

File: thrust/thrust/system/omp/detail/scan.h [hunk 10]
```diff
@@ -117,7 +122,13 @@ OutputIterator scan_impl(
...
```
Comment (dkolsen-pgi) [inline thrust/thrust/system/omp/detail/scan.h]: You added the assertion as suggested, but forgot to remove this now unnecessary check for `num_threads > 1`.

Comment (charan-003) [reply thrust/thrust/system/omp/detail/scan.h]: Right, thanks for bringing this up.
